### PR TITLE
Refactor metasyn and make CLI more feature complete

### LINKED ISSUE #145: Idea: make actual object for var_spec
This would allow us to check before doing MetaFrame.fit_dataset() whether var_spec is correct, and also it would do autocomplete in your IDE!

### PR DESCRIPTION
**Pull request by qubixes**
Major update for metasyn 1.0.

Breaking/major changes (as far as I know):

- In `MetaFrame.from_dataframe()`, `spec` becomes `var_specs` and changes from a dictionary to a list.
- `MetaFrame`: A new `from_config` classmethod to create a datafree MetaFrame.
- CLI interface is now feature complete using .toml files as input for the configuration.
- CLI: For the `create_meta` method, the input and output parameters are now optional.
- `UniformDistribution` has its parameters changed to `low` and `high` to conform with the discrete version.
- `MetaVar`: The `detect` method is gone and you should now either use the `fit` method or the `__init__` itself.
- `MetaVar`: 
- `Privacy`: Privacy classes can and should now be dynamically loaded using entry points.
- `MetaConfig`: A new configuration object that is used to parse and check toml/dictionaries.
- `VarConfig`: A new variable configuration object.
- `VarConfigAccess`: An accessor for the configuration object that takes into account defaults from the meta_config.
- `DistributionSpec`: A new distribution configuration class that can parse and check configurations.
- `DistributionProviderList`: A new `create` method to allow for datafree creation of distributions.
- `fit_kwargs` is now in the `DistributionSpec`
- `unique` is now in the `DistributionSpec`
Fixes #145

### CODE REVIEW

**Comment by qubixes** [comment]
It is currently a bit messy in places (code duplication, etc), so this is for now more a showcase of what functionality can be implemented. Since TOML is a bit more user friendly, I figured that it could also be used to generate synthetic data from scratch, or add new columns that are not in the original dataset. It does need quite a bit of refactoring, before this PR can be merged though.

File: docs/source/usage/cli.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@ Metasyn provides a command-line interface (CLI) for accessing core functionality
 
 The CLI currently has three subcommands:
 
-* The ``create-meta`` subcommand, which allows you to **create generative metadata** from a ``CSV file``
+* The ``create-meta`` subcommand, which allows you to **create generative metadata** from a ``CSV file`` and/or configuration file.
```
Comment (Samuwhale) [inline docs/source/usage/cli.rst:7]:
```suggestion
* The ``create-meta`` subcommand, which allows you to **create generative metadata** from a ``.csv file`` and/or a ``.toml`` configuration file.
```

File: docs/source/usage/generating_metaframes.rst
Comment (Samuwhale) [inline docs/source/usage/generating_metaframes.rst:1]: Is there a reason why sentences are manually separated across lines, as opposed to letting page wrap them automatically?

File: examples/example_config.toml
Comment (Samuwhale) [inline examples/example_config.toml:1]: This looks super clean, nice!

File: metasyn/config.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
+"""Configuration classes for creating metaframes."""
```
Comment (Samuwhale) [inline metasyn/config.py:1]:
```suggestion
"""Module defining configuration classes for creating MetaFrames."""
```

File: metasyn/config.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+except ImportError:
+    import tomli as tomllib  # type: ignore  # noqa
+
+from metasyn.privacy import BasePrivacy, get_privacy
+from metasyn.provider import DistributionProviderList
+from metasyn.util import VarConfig
+
+
+class MetaConfig():
+    """Configuration class for creating MetaFrames.
```
Comment (Samuwhale) [inline metasyn/config.py:18]:
```suggestion
    """
    Manages configurations for individual variables in a MetaFrame.

    This class is used to create, manipulate, and retrieve configurations for
    individual variables in a MetaFrame. It also provides methods for loading
    configurations from .toml files and converting configurations to dictionaries.
```

Feel free to change it to your liking, but I think it's nice to have it be a bit more descriptive from the start.

File: metasyn/util.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
+"""Configuration classes for creating metaframes."""
```
Comment (Samuwhale) [inline metasyn/util.py:1]:
```suggestion
"""Utility module for MetaSyn.

This module provides utility classes that are used across MetaSyn,
including classes for specifying distributions and storing variable
configurations.
"""
```

Not sure if this is a better description, but it's nice to not have it be the exact same as the config.py one

**Review by Samuwhale — COMMENTED** [review]
Overall very nice work. I'd love to hear a bit more about this next Monday, but I like it!

File: docs/source/usage/generating_metaframes.rst
Comment (qubixes) [reply docs/source/usage/generating_metaframes.rst:1]: Because I don't have it wrapped in my editor..! I agree that it is sometimes annoying to have to redo the wrapping manually.

File: examples/example_config.toml
Comment (qubixes) [reply examples/example_config.toml:1]: Thanks!

**Comment by qubixes** [comment]
@Samuwhale Thanks for you comments and suggestions, I have applied them verbatim.

**Review by Samuwhale — APPROVED** [review]

File: metasyn/config.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            privacy = get_privacy(**privacy)
+
+        self.privacy = privacy
+        self.n_rows = n_rows
+
+    @staticmethod
+    def _parse_var_config(var_cfg):
+        if isinstance(var_cfg, VarConfig):
+            return var_cfg
+        return VarConfig.from_dict(var_cfg)
```
Comment (vankesteren) [inline metasyn/config.py:65]: nice

File: metasyn/config.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        self.meta_config = meta_config
+
+    def __getattribute__(self, attr):
+        if attr == "privacy":
+            if self.var_config.privacy is None:
+                return self.meta_config.privacy
+            return self.var_config.privacy
+        if attr not in ("var_config", "meta_config") and hasattr(self.var_config, attr):
+            return getattr(self.var_config, attr)
+        return super().__getattribute__(attr)
```
Comment (vankesteren) [inline metasyn/config.py:175]: I think this is kind of strange. So this is a class around the class just to get the attributes of the class?? Can't we put this in MetaConfig or VarConfig directly?

File: metasyn/metaframe.py [hunk 7]
```diff
@@ -53,11 +53,11 @@ def n_columns(self) -> int:
     @classmethod
     def fit_dataframe(
             cls,
-            df: pl.DataFrame,
-            spec: Optional[dict[str, dict]] = None,
-            dist_providers: Union[str, list[str], BaseDistributionProvider,
-                                  list[BaseDistributionProvider]] = "builtin",
-            privacy: Optional[BasePrivacy] = None,
+            df: Optional[Union[pl.DataFrame, pd.DataFrame]],
```
Comment (vankesteren) [inline metasyn/metaframe.py:56]: So we support either now? I like that but I don't want pandas as a dependency!

This change should also be in docs

File: metasyn/metaframe.py [hunk 8]
```diff
@@ -53,11 +53,11 @@ def n_columns(self) -> int:
...
+            meta_config: Optional[MetaConfig] = None,
+            var_specs: Optional[list[dict]] = None,
```
Comment (vankesteren) [inline metasyn/metaframe.py:58]: why change this from spec to var specs and from dict[str, dict] to list[dict]?

File: metasyn/metaframe.py [hunk 9]
```diff
@@ -120,46 +122,70 @@ def fit_dataframe(
...
-            spec = deepcopy(spec)
+            assert privacy is None
 
-        if set(list(spec)) - set(df.columns):
-            raise ValueError(
-                "Argument 'spec' includes the specifications for column names that do "
-                "not exist in the supplied dataframe:"
-                f" '{set(list(spec)) - set(df.columns)}'")
+        if isinstance(df, pd.DataFrame):
+            df = pl.DataFrame(df)
```
Comment (vankesteren) [inline metasyn/metaframe.py:137]: Maybe change this from `isinstance(df, pd.DataFrame)` to `not isinstance(df, pl.DataFrame)` and then try the conversion. If there is a good error message this will automatically check everything and also support everything that polars supports!

File: metasyn/metaframe.py [hunk 10]
```diff
@@ -120,46 +122,70 @@ def fit_dataframe(
...
-        if spec is None:
-            spec = {}
+        if meta_config is None:
+            if privacy is None:
+                privacy = {"name": "none"}
+            elif isinstance(privacy, BasePrivacy):
+                privacy = privacy.to_dict()
+            var_specs = [] if var_specs is None else var_specs
+            dist_providers = dist_providers if dist_providers is not None else ["builtin"]
+            meta_config = MetaConfig(var_specs, dist_providers, privacy)
```
Comment (vankesteren) [inline metasyn/metaframe.py:132]: My comments above would lead to pretty small changes here I think, and not elsewhere.

File: metasyn/metaframe.py [hunk 11]
```diff
@@ -120,46 +122,70 @@ def fit_dataframe(
...
+            for col_name in tqdm(columns, disable=not progress_bar):
+                var_spec = meta_config.get(col_name)
+                var = MetaVar.fit(
+                    df[col_name],
+                    var_spec.dist_spec,
+                    meta_config.dist_providers,
+                    var_spec.privacy,
+                    var_spec.prop_missing,
+                    var_spec.description)
+                all_vars.append(var)
```
Comment (vankesteren) [inline metasyn/metaframe.py:150]: This feels so nice now. Much better & more legible than before!

File: metasyn/metaframe.py [hunk 12]
```diff
@@ -120,46 +122,70 @@ def fit_dataframe(
...
-
+        if df is None:
+            if meta_config.n_rows is None:
+                raise ValueError("Please provide the number of rows in the configuration, "
+                                 "or supply a DataFrame.")
+            return cls(all_vars, meta_config.n_rows)
         return cls(all_vars, len(df))
 
+    @classmethod
+    def from_config(cls, meta_config: MetaConfig) -> MetaFrame:
```
Comment (vankesteren) [inline metasyn/metaframe.py:175]: 🤩

File: metasyn/metaframe.py [hunk 13]
```diff
@@ -236,8 +262,11 @@ def export(self, fp: Union[pathlib.Path, str],
         self_dict = _jsonify(self.to_dict())
         if validate:
             validate_gmf_dict(self_dict)
-        with open(fp, "w", encoding="utf-8") as f:
-            json.dump(self_dict, f, indent=4)
+        if fp is None:
+            print(json.dumps(self_dict, indent=4))
```
Comment (vankesteren) [inline metasyn/metaframe.py:266]: ah yes nice small change

File: metasyn/util.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+            Dictionary to parse the configuration from.
+
+        Returns
+        -------
+            A new VarConfig instance.
+        """
+        dist_spec = var_dict.pop("distribution", None)
+        if dist_spec is None:
+            return cls(**var_dict)
+        return cls(**var_dict, dist_spec=DistributionSpec.parse(dist_spec))
```
Comment (vankesteren) [inline metasyn/util.py:141]: Ha this is nice small code.

File: metasyn/var.py [hunk 15]
```diff
@@ -47,82 +48,23 @@ class MetaVar():
...
-                 series: Optional[Union[pl.Series, pd.Series]] = None,
-                 name: Optional[str] = None,
-                 distribution: Optional[BaseDistribution] = None,
-                 prop_missing: Optional[float] = None,
-                 dtype: Optional[str] = None,
-                 description: Optional[str] = None):
+                 distribution: BaseDistribution,
+                 dtype: str = "unknown",
+                 description: Optional[str] = None,
+                 prop_missing: float = 1.0):
```
Comment (vankesteren) [inline metasyn/var.py:57]: default 1.0?? is this for "data-free"? otherwise I'd rather expect prop_missing to be 0.0 by default.

File: metasyn/var.py [hunk 16]
```diff
@@ -197,13 +139,14 @@ def __str__(self) -> str:
             f'- Distribution:\n{distribution_formatted}\n'
         )
 
-    def fit(self,  # pylint: disable=too-many-arguments
-            dist: Optional[Union[str, BaseDistribution, type]] = None,
-            dist_providers: Union[str, type,
-                                  BaseDistributionProvider] = "builtin",
+    @classmethod
+    def fit(cls,  # pylint: disable=too-many-arguments
+            series: Union[pl.Series, pd.Series],
```
Comment (vankesteren) [inline metasyn/var.py:144]: again pd.Series support is nice but it would be nice to not have to depend on pd

File: metasyn/var.py
Comment (vankesteren) [inline metasyn/var.py:1]: A lot changed in this file. Might be nice to briefly talk about it.

File: tests/data/no_data_config.toml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+# Example toml file as input for metasyn
+
+[general]
+n_rows = 100
+
+
+[[var]]
+
```
Comment (vankesteren) [inline tests/data/no_data_config.toml:8]:
```suggestion
```

File: tests/data/no_data_config.toml [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+name = "PassengerId"
+data_free = true
+prop_missing = 0.0
+description = "ID of the unfortunate passenger."
+var_type = "discrete"
+distribution = {implements = "core.unique_key", unique = true, parameters = {consecutive = 1, low = 0}}
+
+
+[[var]]
+
```
Comment (vankesteren) [inline tests/data/no_data_config.toml:18]:
```suggestion
```

File: tests/test_cli.py [hunk 19]
```diff
@@ -35,17 +36,20 @@ def tmp_dir(tmp_path_factory) -> Path:
             "Fare": float
         }
         data_frame = pl.read_csv(csv_fp, dtypes=csv_dt)[:100]
-        meta_frame = MetaFrame.fit_dataframe(data_frame, spec={"PassengerId": {"unique": True}})
+        meta_frame = MetaFrame.fit_dataframe(data_frame, var_specs=[{"name": "PassengerId", "distribution": {"unique": True}}])
```
Comment (vankesteren) [inline tests/test_cli.py:39]: Ah now I see why it changed to a list instead of a dict

Comment (vankesteren) [reply tests/test_cli.py:39]: can we still do arbitrary order which does not align with the order of the columns in the data frame?

**Review by vankesteren — COMMENTED** [review]
Great, big work! We'll talk soon, but this seems to be pretty complete. After some minor comments I think we can approve this and make a new release soon.

**Comment by vankesteren** [comment]
NB: for the changelog in the new version we should maybe rename this PR?

File: docs/source/usage/cli.rst [hunk 20]
```diff
@@ -126,17 +126,43 @@ The ``create-meta`` command also takes one optional argument:
...
    An example of a configuration file that specifies the ``PassengerId`` column to be unique and the ``Fare`` column to have a log-normal distribution is as follows:
 
-   .. code-block:: ini
+   .. code-block:: toml
 
-      [var.PassengerId]
-      unique = True
+      [[var]]
+      name = "PassengerId"
+      distribution = {unique = true}  # Notice lower capitalization for .toml files.
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/usage/cli.rst:137]: Python -> True
TOML -> true

That is why I put in the comment there. Are you sure you want to remove it, or perhaps reword it?

File: docs/source/usage/cli.rst [hunk 21]
```diff
@@ -126,17 +126,43 @@ The ``create-meta`` command also takes one optional argument:
...
 
-      [var.Fare]
-      distribution=LogNormalDistribution
+
+      [[var]]
+      name = "Fare"
+      distribution = {implements = "core.log_normal"}
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/usage/cli.rst:142]: Yes, it's possible, but do we want to advise our users to use that? If there is a "extra.log_normal" in a different distribution, then it will be selected depending on the order of the providers..

File: docs/source/usage/generating_metaframes.rst [hunk 22]
```diff
@@ -91,30 +106,29 @@ spec
...
     .. code-block:: python
-        
+
         from metasyn.distribution import FakerDistribution, DiscreteUniformDistribution, RegexDistribution
+        from metasyn.config import VarConfig, DistributionSpec
 
         # Create a specification dictionary for generating synthetic data
-        var_spec = {
-
+        var_specs = [
```
[the remark this answers is not in the document]
Comment (qubixes) [reply docs/source/usage/generating_metaframes.rst:114]: Yes, it is more wordy. However, personally I think it is better and less error-prone this way. So yes, I think we could provide a dictionary, but wasn't a big part of the point that you could get help from your IDE/type checker to make sure that you put it in correctly? Also, I saw that you figured out that we don't do dictionaries anymore below, so it will be more wordy regardless of supplying a list of dicts instead of a list of VarConfig's.

File: metasyn/__main__.py [hunk 23]
```diff
@@ -149,6 +129,10 @@ def synthesize() -> None:
         help="preview six-row synthesized data frame in console and exit",
         action="store_true",
     )
+    parser.add_argument(
+        "-s", "--settings",
```
[the remark this answers is not in the document]
Comment (qubixes) [reply metasyn/__main__.py:133]: It actually didn't do anything at all so I removed it.

File: metasyn/metaframe.py (hunk 7 above)
Comment (qubixes) [reply metasyn/metaframe.py:56]: We have always supported both technically, at least a lot of our test suite checks the results for both. I'm okay with trying to remove pandas as a dependencies and relying on polars to do the heavy lifting. But I think we should do that in a separate PR. Do you agree?

File: metasyn/metaframe.py (hunk 10 above)
Comment (qubixes) [reply metasyn/metaframe.py:132]: I'm not sure which comments you are thinking about here?

File: metasyn/metaframe.py (hunk 9 above)
Comment (qubixes) [reply metasyn/metaframe.py:137]: Makes sense!

File: tests/data/no_data_config.toml (hunk 17 above)
Comment (qubixes) [reply tests/data/no_data_config.toml:8]: I actually wanted the vars to be a little spaced away from the general table, do you disagree?

File: tests/test_cli.py (hunk 19 above)
Comment (qubixes) [reply tests/test_cli.py:39]: Yes, if the column exists, the order doesn't matter. It only matters if the column is a data-free one.

**Comment by qubixes** [comment]
> NB: for the changelog in the new version we should maybe rename this PR?

You mean to make clear that more was done than just the configuration file?

File: docs/source/usage/cli.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (Samuwhale) [reply docs/source/usage/cli.rst:137]: I'd keep the comment but rephrase it to explicitly mention that booleans are lowercase in toml

**Comment by vankesteren** [comment]
Yeah 👍

File: tests/data/no_data_config.toml (hunk 17 above)
Comment (vankesteren) [reply tests/data/no_data_config.toml:8]: nah, fine

File: metasyn/var.py (hunk 16 above)
Comment (vankesteren) [reply metasyn/var.py:144]: will do this as part of #230

File: docs/source/usage/cli.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (vankesteren) [reply docs/source/usage/cli.rst:142]: makes sense to stick with core.log_normal

**Review by vankesteren — APPROVED** [review]
